Issues #11 and #9 - Shuttle bus navigation and changing the view in navigation mode

makchamp opened the pull request:
### #11 Enhancement:

- Finding the closest Shuttle Bus stop using the haversine formula

The haversine formula is used to calculate the distance between two points on a sphere
I use it to find the distances between the starting location and the 2 Concordia shuttle bus stops

- Adding waypoints to ShuttleBus routes and accurate route times 

When the user selects the shuttle transportation method, two waypoints will be set.
The waypoints are determined by which is the nearest shuttle bus stop to the origin location.
For example, if the closest shuttle bus stop is the one located at SGW, it will generate a path from
origin → SGW campus → LOY campus → destination.

- Adding a condition to generate walking instructions when the selected

If the selected mode is the shuttle, we want to show walking instructions towards and from the bus stops.
The instructions when we are inside the shuttle bus will be overridden to display that fact.
A bad side effect of this change is that if the distance is too far to realistically walk to, it will still show walking instructions.
The displayed time is also unaffected by this. Meaning, it will only the show the time of the actual shuttle trip (ignoring the walking) (FIXED, shows more accurate time now)
To remedy this, it is possible to pass the previously calculated distance between the origin and the bus stop to the generateDirections()
We would then check, if the distance is above a certain threshold, we would generate directions using other transportation methods
Whether the chosen method is biking, driving or transit is unclear how that will be decided or if it is even required as only walking directions were requested.
This is due to the assumption that the shuttle option will mostly be picked when in or around Concordia, where walking is practical.

### #9 - Enhancement

- The camera now moves to the view based on the steps

Upon clicking the next and previous step buttons shown on screen, the view of the camera will change to its respective coordinate.

- Enhanced step by step instruction to rotate view based on bearing 

Now each step by step instruction will have their view rotate in such a way that the path is pointed to the north just like we see it on the street view. This is done with the help of getBearing method call which takes 2 points on the map and calculates the bearing between them in order to rotate the camera to the appropriate degree.

- Major refactoring of the generateDirections method

The generateDirections() method has grown to be too large and convoluted.
It required some major refactoring to be able to move forward with custom shuttle instructions (while keeping google pathing and partial instructions).
I have split the method into 4:

 1. fetchDirections() : sends a directions request and retrieves it as a JSON object, calls parseDirections() on response.

 2. parseDirections(): goes through the JSON response to parse and store all directions data. Calls 3 and 4: 

  3. parseInstructions() 

  4. parseCoordinates()
as needed.

This change cuts down a lot of redundant code. Repeating code is now called with functions.
Changed instructions coordinates to be stored directly together as a LatLng object instead of sequentially as strings. I feel like it is more logical to keep lat/lng coordinates tied together.
Modifed the showInstructions method accordingly to handle direct coordinates instead of strings. Also, I modifed the NavigationRoute class to accommodate this change.

### Optimized imports

- Remove the *Navigate with shuttle button* from the ShuttleFragment 

- Renamed the *Shuttle Bus* menu item to *Shuttle Bus Schedule*

- Added markers for the stops  (can be changed removed if too distracting)

Amanda-Kolopa commented on app/src/androidTest/java/com/droidhats/campuscompass/views/ShuttleBusTest.kt:
```diff
@@ -159,12 +151,6 @@ class ShuttleBusTest {
-                withId(R.id.navigateWithShuttle), withText("Start Shuttle Trip"),
-                isDisplayed())).perform(click())
-
         //Click to view LOY campus view
         onView(
             allOf(
```
From this point on, please remove the checks for the toggle button.
This is causing the test to fail.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/models/Location.kt:
```diff
@@ -5,11 +5,12 @@ import com.droidhats.campuscompass.helpers.Observer
 import com.google.android.libraries.places.api.model.Place
 import kotlinx.android.parcel.Parcelize
 
+
```
Unnecessary extra line. Please remove it.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/models/Map.kt:
```diff
@@ -46,6 +47,17 @@ class Map(
         }
         setBuildingMarker()
+        addShuttleStopMarkers()
+
```
Unnecessary extra line. Please remove it.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/repositories/NavigationRepository.kt:
```diff
@@ -114,7 +126,23 @@ class NavigationRepository(private val application: Application) {
+                        }
+                        else
+                            times[method.string] = legs.getJSONObject("duration").getString("text")
+
```
Unnecessary extra line. Please remove it.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -92,14 +100,14 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
     private fun sendSQLiteQuery(query: String): Boolean {
-     if (query.isBlank()) return false
-     val qEsc = query.replace("'","")
+        if (query.isBlank()) return false
+        val qEsc = query.replace("'","")
```
Can these be changed to tabbed spaces, please?

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -92,14 +100,14 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
-             "WHERE location_type ='classroom' " +
-              "AND location_name like '%$qEsc%' " +
-               "OR location_name like '%${qEsc.toUpperCase(Locale.ROOT)}%' " +
+                    "FROM IndoorLocation " +
+                    "WHERE location_type ='classroom' " +
+                    "AND location_name like '%$qEsc%' " +
+                    "OR location_name like '%${qEsc.toUpperCase(Locale.ROOT)}%' " +
```
Can these be changed to tabbed indentations, please?

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -118,20 +126,59 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
-               navigationRepository.fetchPlace(origin)
-           if (destination is GooglePlace && !destination.isCurrentLocation)
-               navigationRepository.fetchPlace(destination)
+        GlobalScope.launch(Dispatchers.Default + handler) {
+            if (origin is GooglePlace && !origin.isCurrentLocation)
+                navigationRepository.fetchPlace(origin)
+            if (destination is GooglePlace && !destination.isCurrentLocation)
+                navigationRepository.fetchPlace(destination)
```
Please remove the extra space for each line.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -118,20 +126,59 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
-           if (destination is Building)
-               navigationRepository.fetchRouteTimes(origin, destination)
-       }
+            if (destination is Building)
+                navigationRepository.fetchRouteTimes(origin, destination, closestShuttle)
+        }
```
Please remove the extra space for each line.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -118,20 +126,59 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
+           NavigationRepository.LOY_TO_SGW_WAYPOINT
+    }
+
+
```
Please remove the extra line.

Amanda-Kolopa commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -118,20 +126,59 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
+    /**
+     * Method to find the closest Concordia shuttle bus stop from a given coordinate
+     */
+
```
Please remove the extra line.

JeremiahTee commented on app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:
```diff
@@ -118,20 +127,61 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
+    /**
+     * Method used to find the distance between two world coordinates using the haversine formula
+     */
+    fun haversine(location1 : Location, location2 : Location) : Double{
```
Make this function private as it's only used in the closestShuttleStop() method

JeremiahTee commented on app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:
```diff
@@ -31,45 +29,25 @@ import com.droidhats.campuscompass.MainActivity
-import com.droidhats.campuscompass.models.GooglePlace
-import com.droidhats.campuscompass.models.IndoorLocation
-import com.droidhats.campuscompass.models.CalendarEvent
+import com.droidhats.campuscompass.models.*
```
Was this intentional or was it done automatically by the IDE? If it wasn't intentional, I would fix these imports to be consistent through all classes

JeremiahTee commented on app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:
```diff
@@ -31,45 +29,25 @@ import com.droidhats.campuscompass.MainActivity
-import com.google.android.gms.location.LocationResult
-import com.google.android.gms.location.LocationServices
-import com.google.android.gms.location.LocationSettingsRequest
+import com.google.android.gms.location.*
```
Same comment as previous one

JeremiahTee commented on app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:
```diff
@@ -31,45 +29,25 @@ import com.droidhats.campuscompass.MainActivity
-import com.google.android.gms.maps.model.PolylineOptions
-import com.google.android.gms.maps.model.Polyline
-import com.google.android.gms.maps.model.CameraPosition
-import com.google.android.gms.maps.model.MapStyleOptions
```
Same comment as previous one

JeremiahTee commented on app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:
```diff
@@ -31,45 +29,25 @@ import com.droidhats.campuscompass.MainActivity
-import kotlin.collections.List
-import kotlin.collections.MutableList
-import kotlinx.android.synthetic.main.bottom_sheet_layout.bottom_sheet
+import kotlinx.android.synthetic.main.bottom_sheet_layout.*
```
Same comment as previous one

makchamp commented:
@Amanda-Kolopa 
`I think the markers should only be displayed on the map once we have selected Shuttle Bus as a mode of transportation. Aside from that, I think they're pretty`

JT updated the markers  f1a2b2ee7b8f087112616e90f8f10682f539b2cc so they disappear when zoomed out (like the building markers) 

`The bottom sheet for instructions should be scrollable. In the following picture, there's text but I can't access it. Also, is it possible to write "Shuttle bus stop will be on the left" instead of Destination? (not a big issue)`

The bottom sheet is now scrollable. That text is also given from google with regular transit instructions. We could change it but it doesn't seem awfully necessary like you said.

`There's an issue when setting the mode of transportation as shuttle and setting a start or end point that is closer to a certain campus.`

Fixed with commits 2ba239156af91d8fa5b16eb541359c2c07d573a8 and 56affc378f46e84ee6002afa53141100be18fc48

`testShuttleBus: You mentioned that you removed the navigation button from the ShuttleFragment. Please remove the specified lines (I left a comment for it).`

Fixed with commit 95cc83ec8b36a1383a124dad124746139153432b
For the other tests, we could look at fixing them after all navigation chaining is completed
